src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,251 @@
+        {
+            UnspentOutputs prevUtxo = coins.UnspentOutputs[0];
+            if (prevUtxo == null)
+            {
+                this.Logger.LogTrace("(-)[PREV_UTXO_IS_NULL]");
+                ConsensusErrors.ReadTxPrevFailed.Throw();
+            }
+
+            return prevUtxo;
+        }
+
+        /// <summary>
+        /// Checks if coinstake transaction is valid.
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        private void CheckIfCoinstakeIsTrue(ProvenBlockHeader header)
+        {
+            if (header.Coinstake.IsCoinStake)
+                return;
+
+            this.Logger.LogTrace("(-)[PROVEN_HEADER_INVALID_MERKLE_PROOF_SIZE]");
+            ConsensusErrors.NonCoinstake.Throw();
+        }
+
+        /// <summary>
+        /// Checks whether header time is equal to the timestamp of the coinstake tx and if coinstake tx
+        /// timestamp is divisible by 16 (using timestamp mask).
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        private void CheckHeaderAndCoinstakeTimes(ProvenBlockHeader header)
+        {
+            uint coinstakeTime = header.Coinstake.Time;
+            uint headerTime = header.Time; 
+
+            // Check if times are equal and coinstake tx time is divisible by 16.
+            if ((headerTime == coinstakeTime) && ((coinstakeTime & PosConsensusOptions.StakeTimestampMask) == 0))
+                return;
+
+            this.Logger.LogTrace("(-)[BAD_TIME]");
+            ConsensusErrors.StakeTimeViolation.Throw();
+        }
+
+        /// <summary>
+        /// Checks the coinstake age requirement.
+        /// </summary>
+        /// <param name="chainedHeader">The chained header.</param>
+        /// <param name="unspentOutputs">The unspent outputs.</param>
+        private void CheckCoinstakeAgeRequirement(ChainedHeader chainedHeader, UnspentOutputs unspentOutputs)
+        {
+            ChainedHeader prevChainedHeader = chainedHeader.Previous;
+
+            var options = (PosConsensusOptions)this.PosParent.Network.Consensus.Options;
+            int targetDepth = options.GetStakeMinConfirmations(prevChainedHeader.Height + 1, this.PosParent.Network) - 1;
+
+            int actualDepth = prevChainedHeader.Height - (int)unspentOutputs.Height;
+            if (actualDepth < targetDepth)
+                return;
+
+            this.Logger.LogTrace("(-)[BAD_STAKE_DEPTH]");
+            ConsensusErrors.InvalidStakeDepth.Throw();
+        }
+
+        /// <see cref="IStakeValidator.VerifySignature"/>
+        private void CheckSignature(ProvenBlockHeader header, UnspentOutputs unspentOutputs)
+        {
+            if (this.stakeValidator.VerifySignature(unspentOutputs, header.Coinstake, 0, ScriptVerify.None))
+                return;
+            
+            this.Logger.LogTrace("(-)[BAD_SIGNATURE]");
+            ConsensusErrors.CoinstakeVerifySignatureFailed.Throw();
```

Comment:
for all methods in that rule that can throw add exception xml doc

Reply:
Done